#               Inpatient Analysis & Predicting Length of Stay
##                                Data Science Project Proposal


#### Project prepared by: Diaa Aldein Alsayed Ibrahim Osman
#### Prepared for: Epsilon AI Institute


Data obtained from https://www.kaggle.com/datasets/jonasalmeida/2015-deidentified-ny-inpatient-discharge-sparcs/data

## About Dataset
Public Health Data
This is the public dataset made available at https://health.data.ny.gov/Health/Hospital-Inpatient-Discharges-SPARCS-De-Identified/82xm-y6g8 by the Dept of Health of New York state. The following description can be found at that page:

* The Statewide Planning and Research Cooperative System (SPARCS) Inpatient De-identified File contains discharge level detail on patient characteristics, diagnoses, treatments, services, and charges. This data file contains basic record level detail for the discharge. The de-identified data file does not contain data that is protected health information (PHI) under HIPAA. The health information is not individually identifiable; all data elements considered identifiable have been redacted. For example, the direct identifiers regarding a date have the day and month portion of the date removed.
---
* The aim of this project is to conduct inpatient analysis and predict the length of stay in the hospital using the parameters likely to be available when the patient is admitted.
---
* The data is unclean, has missing values, and contains 2.35 million rows and 37 columns. It may not be necessary to include all instances and features to achieve the goal of this project.

## Data Cleaning & preparation Step:

In [1]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import warnings            
warnings.filterwarnings("ignore") 

In [2]:
# loading the dataset.
df = pd.read_csv('../data/raw/Hospital_Inpatient_Discharges__SPARCS_De-Identified___2015.csv',low_memory=False)

In [3]:
# veiwing the head of the data the first 5 rows
df.head()

Health Service Area Hospital County  Operating Certificate Number  \
0          Western NY     Cattaraugus                      401001.0   
1          Western NY     Cattaraugus                      401001.0   
2          Western NY     Cattaraugus                      401001.0   
3          Western NY     Cattaraugus                      401001.0   
4          Western NY      Chautauqua                      602001.0   

   Facility Id                  Facility Name Age Group Zip Code - 3 digits  \
0         66.0         Olean General Hospital  50 to 69                 147   
1         66.0         Olean General Hospital  18 to 29                 147   
2         66.0         Olean General Hospital  30 to 49                 147   
3         66.0         Olean General Hospital  50 to 69                 147   
4        103.0  Woman's Christian Association  50 to 69                 147   

  Gender   Race          Ethnicity  ... Payment Typology 2 Payment Typology 3  \
0      F  White  Not Span/Hispanic  ...                NaN                NaN   
1      F  White  Not Span/Hispanic  ...                NaN                NaN   
2      M  White  Not Span/Hispanic  ...                NaN                NaN   
3      F  White  Not Span/Hispanic  ...                NaN                NaN   
4      M  White  Not Span/Hispanic  ...                NaN                NaN   

  Attending Provider License Number  Operating Provider License Number  \
0                          267162.0                                NaN   
1                          275408.0                                NaN   
2                          267162.0                                NaN   
3                          271326.0                                NaN   
4                          243034.0                                NaN   

   Other Provider License Number Birth Weight  Abortion Edit Indicator  \
0                            NaN            0                        N   
1                            NaN            0                        N   
2                            NaN            0                        N   
3                            NaN            0                        N   
4                            NaN            0                        N   

  Emergency Department Indicator  Total Charges Total Costs  
0                              Y       $5620.00    $4561.28  
1                              Y       $7404.18    $6933.94  
2                              Y       $7931.28    $7154.85  
3                              Y      $11304.44    $4584.11  
4                              Y       $8330.50    $4361.74  

[5 rows x 37 columns]

In [4]:
# Geitting Information about the Data
df.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2346760 entries, 0 to 2346759
Data columns (total 37 columns):
 #   Column                               Non-Null Count    Dtype  
---  ------                               --------------    -----  
 0   Health Service Area                  2343849 non-null  object 
 1   Hospital County                      2343849 non-null  object 
 2   Operating Certificate Number         2343849 non-null  float64
 3   Facility Id                          2343849 non-null  float64
 4   Facility Name                        2346760 non-null  object 
 5   Age Group                            2346760 non-null  object 
 6   Zip Code - 3 digits                  2342333 non-null  object 
 7   Gender                               2346760 non-null  object 
 8   Race                                 2346760 non-null  object 
 9   Ethnicity                            2346760 non-null  object 
 10  Length of Stay                       2346760 non-null  object 
 11

In [5]:
# Modifing columns name
df.columns = df.columns.str.strip().str.lower().str.replace(" ","_").str.replace("-","")
df.columns

Index(['health_service_area', 'hospital_county',
       'operating_certificate_number', 'facility_id', 'facility_name',
       'age_group', 'zip_code__3_digits', 'gender', 'race', 'ethnicity',
       'length_of_stay', 'type_of_admission', 'patient_disposition',
       'discharge_year', 'ccs_diagnosis_code', 'ccs_diagnosis_description',
       'ccs_procedure_code', 'ccs_procedure_description', 'apr_drg_code',
       'apr_drg_description', 'apr_mdc_code', 'apr_mdc_description',
       'apr_severity_of_illness_code', 'apr_severity_of_illness_description',
       'apr_risk_of_mortality', 'apr_medical_surgical_description',
       'payment_typology_1', 'payment_typology_2', 'payment_typology_3',
       'attending_provider_license_number',
       'operating_provider_license_number', 'other_provider_license_number',
       'birth_weight', 'abortion_edit_indicator',
       'emergency_department_indicator', 'total_charges', 'total_costs'],
      dtype='object')

In [6]:
#checking for missing data percentage
df.isnull().mean()*100

health_service_area                     0.124043
hospital_county                         0.124043
operating_certificate_number            0.124043
facility_id                             0.124043
facility_name                           0.000000
age_group                               0.000000
zip_code__3_digits                      0.188643
gender                                  0.000000
race                                    0.000000
ethnicity                               0.000000
length_of_stay                          0.000000
type_of_admission                       0.000000
patient_disposition                     0.000000
discharge_year                          0.000000
ccs_diagnosis_code                      0.000000
ccs_diagnosis_description               0.000000
ccs_procedure_code                      0.000000
ccs_procedure_description               0.000000
apr_drg_code                            0.000000
apr_drg_description                     0.000000
apr_mdc_code        

* As the data contains 2.35 million rows and 37 columns then For any columns that have more than 20% missing values we are going to drop it. Affected columns as follows:  
1. payment_typology_2 by                     32.490264 %
2. payment_typology_3 by                    70.124635 %
3. operating_provider_license_number by     26.114227 %
4. other_provider_license_number by         96.957239 %
---
* For less than 20 % we are going to drop rows. affected columns as follows:
1. health_service_area by                     0.124166 %
2. hospital_county  by                        0.124166 %
3. operating_certificate_number by            0.124166 %
4. facility_id by                             0.124166 %
5. zip_code__3_digits by                      0.188744 %
6. apr_severity_of_illness_description by     0.004777 %
7. apr_risk_of_mortality by                   0.004777 %
8. attending_provider_license_number by       0.124166 %

In [7]:
df.drop(['payment_typology_2','payment_typology_3','operating_provider_license_number','other_provider_license_number'],axis=1,inplace=True)

In [8]:
df.dropna(axis=0,inplace=True)

In [9]:
#checking for missing data percentage
df.isnull().mean()*100

health_service_area                    0.0
hospital_county                        0.0
operating_certificate_number           0.0
facility_id                            0.0
facility_name                          0.0
age_group                              0.0
zip_code__3_digits                     0.0
gender                                 0.0
race                                   0.0
ethnicity                              0.0
length_of_stay                         0.0
type_of_admission                      0.0
patient_disposition                    0.0
discharge_year                         0.0
ccs_diagnosis_code                     0.0
ccs_diagnosis_description              0.0
ccs_procedure_code                     0.0
ccs_procedure_description              0.0
apr_drg_code                           0.0
apr_drg_description                    0.0
apr_mdc_code                           0.0
apr_mdc_description                    0.0
apr_severity_of_illness_code           0.0
apr_severit

In [10]:
pd.set_option('display.max_columns', None)  # This will display all columns

In [11]:
df.head()

health_service_area hospital_county  operating_certificate_number  \
0          Western NY     Cattaraugus                      401001.0   
1          Western NY     Cattaraugus                      401001.0   
2          Western NY     Cattaraugus                      401001.0   
3          Western NY     Cattaraugus                      401001.0   
4          Western NY      Chautauqua                      602001.0   

   facility_id                  facility_name age_group zip_code__3_digits  \
0         66.0         Olean General Hospital  50 to 69                147   
1         66.0         Olean General Hospital  18 to 29                147   
2         66.0         Olean General Hospital  30 to 49                147   
3         66.0         Olean General Hospital  50 to 69                147   
4        103.0  Woman's Christian Association  50 to 69                147   

  gender   race          ethnicity length_of_stay type_of_admission  \
0      F  White  Not Span/Hispanic              5         Emergency   
1      F  White  Not Span/Hispanic              8         Emergency   
2      M  White  Not Span/Hispanic              8         Emergency   
3      F  White  Not Span/Hispanic              4         Emergency   
4      M  White  Not Span/Hispanic              3         Emergency   

  patient_disposition  discharge_year  ccs_diagnosis_code  \
0   Home or Self Care            2015                 657   
1   Home or Self Care            2015                 657   
2   Home or Self Care            2015                 657   
3   Home or Self Care            2015                 128   
4   Home or Self Care            2015                  83   

  ccs_diagnosis_description  ccs_procedure_code ccs_procedure_description  \
0            Mood disorders                   0                   NO PROC   
1            Mood disorders                   0                   NO PROC   
2            Mood disorders                   0                   NO PROC   
3                    Asthma                   0                   NO PROC   
4     Epilepsy; convulsions                   0                   NO PROC   

   apr_drg_code apr_drg_description  apr_mdc_code  \
0           753   Bipolar disorders            19   
1           753   Bipolar disorders            19   
2           753   Bipolar disorders            19   
3           141              Asthma             4   
4            53             Seizure             1   

                                apr_mdc_description  \
0                     Mental Diseases and Disorders   
1                     Mental Diseases and Disorders   
2                     Mental Diseases and Disorders   
3  Diseases and Disorders of the Respiratory System   
4      Diseases and Disorders of the Nervous System   

   apr_severity_of_illness_code apr_severity_of_illness_description  \
0                             1                               Minor   
1                             2                            Moderate   
2                             1                               Minor   
3                             1                               Minor   
4                             2                            Moderate   

  apr_risk_of_mortality apr_medical_surgical_description payment_typology_1  \
0                 Minor                          Medical           Medicaid   
1                 Minor                          Medical           Medicaid   
2                 Minor                          Medical           Medicaid   
3                 Minor                          Medical           Medicare   
4              Moderate                          Medical           Medicaid   

   attending_provider_license_number  birth_weight abortion_edit_indicator  \
0                           267162.0             0                       N   
1                           275408.0             0                       N   
2                           267162.0             0             

In [12]:
# checking features unique values.
for i in list(df.columns):
    print(i)
    print(df[i].unique())
    print("*"*100)

health_service_area
['Western NY' 'Finger Lakes' 'Central NY' 'Southern Tier' 'Capital/Adiron'
 'Hudson Valley' 'New York City' 'Long Island']
****************************************************************************************************
hospital_county
['Cattaraugus' 'Chautauqua' 'Erie' 'Genesee' 'Niagara' 'Chemung'
 'Livingston' 'Monroe' 'Ontario' 'Steuben' 'Wayne' 'Yates' 'Oswego'
 'Broome' 'Cayuga' 'Chenango' 'Cortland' 'Herkimer' 'Jefferson' 'Lewis'
 'Madison' 'Oneida' 'Onondaga' 'St Lawrence' 'Tompkins' 'Albany' 'Clinton'
 'Columbia' 'Franklin' 'Fulton' 'Montgomery' 'Otsego' 'Rensselaer'
 'Saratoga' 'Schenectady' 'Warren' 'Dutchess' 'Orange' 'Putnam' 'Rockland'
 'Sullivan' 'Ulster' 'Westchester' 'Bronx' 'Kings' 'Manhattan' 'Richmond'
 'Queens' 'Suffolk' 'Nassau' 'Allegany' 'Wyoming' 'Orleans' 'Schuyler'
 'Delaware' 'Essex' 'Schoharie']
****************************************************************************************************
operating_certificate_number
[ 401001. 

['50 to 69' '18 to 29' '30 to 49' '70 or Older' '0 to 17']
****************************************************************************************************
zip_code__3_digits
['147' '141' '142' '143' '140' 'OOS' '148' '146' '145' '144' '149' '130'
 '131' '139' '137' '138' '136' '117' '133' '134' '132' '135' '123' '128'
 '122' '121' '120' '124' '104' '125' '129' '116' '110' '126' '109' '127'
 '105' '106' '108' '100' '112' '107' '114' '113' '111' '103' '115' '119'
 '101' '118']
****************************************************************************************************
gender
['F' 'M' 'U']
****************************************************************************************************
race
['White' 'Other Race' 'Black/African American' 'Multi-racial']
****************************************************************************************************
ethnicity
['Not Span/Hispanic' 'Unknown' 'Spanish/Hispanic' 'Multi-ethnic']
***********************************************

['NO PROC' 'RESP INTUB/MECH VENTIL' 'OT DX PRC (INTERVW,EVAL'
 'TRACTN, SPLNT, OT WOUND' 'OTHER THERAPEUTIC PRCS'
 'OT PRCS TO ASSIST DELIV' 'CESAREAN SECTION' 'ELECTROGR CARDIAC MONIT'
 'BIOPSY OF LIVER' 'BLOOD TRANSFUSION' 'REPAIR CUR OBS LACERATN'
 'PSYCHO/PSYCHI EVAL/THER' 'SUTURE SKIN/SUBCUT TISS' 'FETAL MONITORING'
 'COLONOSCOPY AND BIOPSY' 'OT NON-OR THER PRC NOSE'
 'DEBRIDE WOUND,INF,BURN' 'UP GASTRO ENDOSC/BIOPSY' 'CT SCAN CHEST'
 'HEMORRHOID PRCS' 'DX ULTRASOUND HEART' 'EPISIOTOMY'
 'OT NON-OR THER PRC; FEM' 'PERIPH VASCULAR BYPASS'
 'DIAGNOSTIC SPINAL TAP' 'OTHER RESP THERAPY' 'SKIN GRAFT' 'APPENDECTOMY'
 'CT SCAN ABDOMEN' 'MAG RESONANCE IMAGING' 'OTHER BOWEL DX PRCS'
 'ART RUPT MEMB ASSIS DLV' 'LAMINECTOMY; EXC IV DSC'
 'FORCEP;VAC;BREACH DELIV' 'ELECTROENCEPHALOGRAM'
 'OT NON-OR LW GI THER PR' 'OT FRACTURE/DISLOC PRC'
 'DX CARDIAC CATHETERIZTN' 'ALCO/DRUG REHAB/DETOX'
 'PROPHYLACTIC VAC/INOCUL' 'OT OR PRCS VES NOT HEAD'
 'COLORECTAL RESECTION' 'OT THER PRCS ON MUSCLES'
 'R

['Mental Diseases and Disorders'
 'Diseases and Disorders of the Respiratory System'
 'Diseases and Disorders of the Nervous System'
 'Diseases and Disorders of the Circulatory System'
 'Diseases and Disorders of the Kidney and Urinary Tract'
 'Diseases and Disorders of the Digestive System'
 'Endocrine, Nutritional and Metabolic Diseases and Disorders'
 'Pregnancy, Childbirth and the Puerperium' 'Burns'
 'Infectious and Parasitic Diseases, Systemic or Unspecified Sites'
 'Diseases and Disorders of the Skin, Subcutaneous Tissue and Breast'
 'Human Immunodeficiency Virus Infections'
 'Diseases and Disorders of the Hepatobiliary System and Pancreas'
 'Alcohol/Drug Use and Alcohol/Drug Induced Organic Mental Disorders'
 'Multiple Significant Trauma'
 'Newborns and Other Neonates with Conditions Originating in the Perinatal Period'
 'Lymphatic, Hematopoietic, Other Malignancies, Chemotherapy and Radiotherapy'
 'Diseases and Disorders of the Eye'
 'Diseases and Disorders of the Female Repro

In [13]:
# checking features value counts.
for i in list(df.columns):
    print(i)
    print(df[i].value_counts())
    print("*"*100)

health_service_area
New York City     1090826
Long Island        339240
Hudson Valley      245917
Capital/Adiron     167322
Western NY         163893
Central NY         158296
Finger Lakes       146558
Southern Tier       30169
Name: health_service_area, dtype: int64
****************************************************************************************************
hospital_county
Manhattan      399693
Kings          248927
Queens         197173
Bronx          188124
Nassau         181142
Suffolk        158098
Erie           120556
Westchester    119375
Monroe         105154
Onondaga        78374
Albany          65717
Richmond        56909
Orange          39500
Dutchess        32594
Rockland        31940
Oneida          29902
Broome          28525
Schenectady     22325
Niagara         17152
Warren          14090
Chemung         13845
Otsego          12265
Rensselaer      11957
Ontario         11950
Ulster          11359
St Lawrence     11106
Jefferson       10823
Chautauqua      10489

0      609688
231    175270
137     83291
228     80961
134     74687
        ...  
126        13
22         13
182         9
46          8
68          7
Name: ccs_procedure_code, Length: 232, dtype: int64
****************************************************************************************************
ccs_procedure_description
NO PROC                    609688
OTHER THERAPEUTIC PRCS     175270
OT PRCS TO ASSIST DELIV     83291
PROPHYLACTIC VAC/INOCUL     80961
CESAREAN SECTION            74687
                            ...  
ABORTION (TERM OF PREG)        13
TYMPANOPLASTY                  13
MAMMOGRAPHY                     9
CORONARY THROMBOLYSIS           8
INJ/LIG ESOPH VARICES           7
Name: ccs_procedure_description, Length: 232, dtype: int64
****************************************************************************************************
apr_drg_code
640    197816
560    146636
720     94929
540     75940
194     56350
        ...  
564        91
583        53
841     

$9973.90      1325
$8375.20       929
$4285.44       814
$4376.42       644
$10575.90      616
              ... 
$20055.23        1
$15700.40        1
$24984.90        1
$32199.70        1
$109189.00       1
Name: total_charges, Length: 1784798, dtype: int64
****************************************************************************************************
total_costs
$1046.14     1324
$1712.33      930
$3061.21      815
$4216.68      646
$1146.07      602
             ... 
$14179.73       1
$74364.83       1
$35281.75       1
$10459.79       1
$1862.39        1
Name: total_costs, Length: 1526541, dtype: int64
****************************************************************************************************


From the above info, unique, and value counts we found that:
1. zip_code__3_digits feature have strange value 'OOS' repeated 67135 times which refer to Out of State zip codes we can replace it with 005 to indicate out of state. Also, changing dtype to numeric.
2. gender feature have strange value 'U' repeated 39 times we are going to drop it.
3. length_of_stay feature have '120 +' value need to be modified, by adding new feature for over_120_stay, then remove + sine. also dtype need to be change to numerical.
4. discharge_year feature have only 2015 value. it will not be useful we are going to drop it.
5. attending_provider_license_number feature it is not useful and we are going to drop it.
6. abortion_edit_indicator feature have only 'N' value. it will not be useful we are going to drop it.
7. total_charges feature have Dollar signe. need to be removed and replace dtype.
8. total_costs feature have Dollar signe. need to be removed and replace dtype.

In [14]:
# starting with zip_code__3_digits feature replacing 'OOS' values to '005' and fixing dtype.
df['zip_code__3_digits'] = df['zip_code__3_digits'].apply(lambda x: str(x).replace('OOS','005'))
df['zip_code__3_digits'] = pd.to_numeric(df['zip_code__3_digits'])
df['zip_code__3_digits'].value_counts()

112    313540
104    224984
117    182027
100    170124
113    121511
114     91275
115     88662
5       67141
142     66140
103     62228
105     61846
109     59262
146     56204
125     46284
140     43809
130     40062
120     34926
119     34085
145     31454
107     30946
121     30480
132     29326
144     29246
110     26779
136     26071
148     25026
128     24437
134     21468
141     21115
124     20468
111     20023
131     18518
123     18445
116     18444
122     17873
147     17279
133     16187
129     15133
137     14450
138     13706
127     12122
126     10865
135     10669
143     10160
108      9388
139      9104
106      9004
118      7932
149      6488
101      5505
Name: zip_code__3_digits, dtype: int64

In [15]:
# droping 'U' values from gender feature
df = df.loc[df.gender != 'U']

In [16]:
df.gender.value_counts()

F    1303589
M    1038593
Name: gender, dtype: int64

In [17]:
# finding out how many values in the data have '120 +' values
df[df['length_of_stay']== '120 +']['length_of_stay'].count()

1857

In [18]:
# function to conver 120 + length of stay to new columns over_120_stay contain (0=false, 1=true)
def over_120_stay(x):
    if x == '120 +':
        return 1
    else:
        return 0
    
df['over_120_stay'] = df['length_of_stay'].apply(over_120_stay)
df['over_120_stay'].value_counts()

0    2340325
1       1857
Name: over_120_stay, dtype: int64

In [19]:
# length_of_stay feature have '120 +' value be modified by remove + sine. also dtype need to be change to numerical.
df['length_of_stay'] = df['length_of_stay'].replace('120 +','120')

In [20]:
df['length_of_stay'].unique()

array(['5', '8', '4', '3', '2', '9', '1', '59', '10', '17', '35', '15',
       '12', '7', '98', '26', '6', '46', '32', '16', '21', '22', '13',
       '77', '14', '58', '28', '11', '18', '19', '29', '27', '42', '24',
       '23', '31', '37', '57', '25', '20', '30', '63', '43', '49', '45',
       '102', '56', '33', '40', '34', '55', '100', '47', '38', '61', '53',
       '117', '62', '120', '80', '48', '70', '44', '51', '41', '64', '50',
       '36', '107', '68', '74', '39', '87', '83', '54', '105', '81',
       '112', '76', '91', '86', '66', '72', '94', '93', '67', '69', '73',
       '119', '115', '75', '97', '92', '95', '60', '65', '82', '79', '71',
       '84', '109', '103', '118', '52', '101', '85', '110', '106', '88',
       '111', '104', '89', '78', '116', '90', '113', '108', '114', '96',
       '99'], dtype=object)

In [21]:
# changing data type to numerical
df['length_of_stay']=pd.to_numeric(df['length_of_stay'])

In [22]:
# droping unecessery features ['discharge_year','attending_provider_license_number','abortion_edit_indicator']
df.drop(['discharge_year','attending_provider_license_number','abortion_edit_indicator'],axis=1,inplace=True)

In [23]:
# replacing Dollar signs in total_charges & total_costs features and changing there dtype from object to numerical
df['total_costs'] = df['total_costs'].str.replace('$','')
df['total_charges'] = df['total_charges'].str.replace('$','')

# changing data type from object to numerical
df['total_costs'] = pd.to_numeric(df['total_costs'])
df['total_charges'] = pd.to_numeric(df['total_charges'])


In [24]:
df['total_costs']

0          4561.28
1          6933.94
2          7154.85
3          4584.11
4          4361.74
            ...   
2346755    6637.09
2346756    3236.18
2346757    1862.39
2346758    2744.13
2346759    2128.89
Name: total_costs, Length: 2342182, dtype: float64

In [25]:
df.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2342182 entries, 0 to 2346759
Data columns (total 31 columns):
 #   Column                               Non-Null Count    Dtype  
---  ------                               --------------    -----  
 0   health_service_area                  2342182 non-null  object 
 1   hospital_county                      2342182 non-null  object 
 2   operating_certificate_number         2342182 non-null  float64
 3   facility_id                          2342182 non-null  float64
 4   facility_name                        2342182 non-null  object 
 5   age_group                            2342182 non-null  object 
 6   zip_code__3_digits                   2342182 non-null  int64  
 7   gender                               2342182 non-null  object 
 8   race                                 2342182 non-null  object 
 9   ethnicity                            2342182 non-null  object 
 10  length_of_stay                       2342182 non-null  int64  
 11

In [28]:
#checking for duplication in data 
df.duplicated().sum()

10598

In [29]:
# Drop duplicates and reset the index
df = df.drop_duplicates().reset_index(drop=True)

In [30]:
df.duplicated().sum()

0

In [31]:
# saving file after cleaning process.
df.to_csv('../data/processed/cleaned_data.csv',index=False)

Data Cleaning & prepration Summery:
* Modifying columns names to be more suitable.
---
* As the data contains 2.35 million rows and 37 columns then For any columns that have more than 20% missing values we drop the columns. Affected columns as follows:  
1. payment_typology_2 by                     32.490264 %
2. payment_typology_3 by                    70.124635 %
3. operating_provider_license_number by     26.114227 %
4. other_provider_license_number by         96.957239 %
.
* For less than 20 % we drop rows. affected columns as follows:
1. health_service_area by                     0.124166 %
2. hospital_county  by                        0.124166 %
3. operating_certificate_number by            0.124166 %
4. facility_id by                             0.124166 %
5. zip_code__3_digits by                      0.188744 %
6. apr_severity_of_illness_description by     0.004777 %
7. apr_risk_of_mortality by                   0.004777 %
8. attending_provider_license_number by       0.124166 %
---
* After checking info, unique, and value counts we found that:

1. zip_code__3_digits feature has strange value 'OOS' repeated 67135 times which refer to Out of State zip codes replaced it with 005 to indicate out of state. Also, changing dtype to numeric.
2. gender feature have strange value 'U' repeated 39 times these values raws dropped.
3. length_of_stay feature has '120 +', we create new feature for over_120_stay then I treated all entries '120 +' in length of stay as 120. also dtype changed to numerical.
4. discharge_year feature has only 2015 value. it will not be useful so drop it.
5. attending_provider_license_number feature it is not useful and we drop it.
6. abortion_edit_indicator feature have only 'N' value. it will not be useful we drop it.
7. total_charges feature have Dollar signe. Sign removed and replaced dtype to numerical.
8. total_costs feature have Dollar signe. Sign removed and replaced dtype to numerical.
---
* Finding 10,598 duplicated record the dealing with them done by dropping them and reset the index.
---
* Saving file as cleaned_data.csv after cleaning process for next Analysis steps.